New Lint: `stacked_ifs`

ARandomDev99 opened the pull request:
Closes #12483 

```
changelog: new lint: [`stacked_ifs`]
```

Alexendoo commented on tests/ui/stacked_ifs.rs:
```diff
@@ -0,0 +1,53 @@
+    } else {
+        println!("False!");
+    }
+}
```
A test for `if if let ...` would be good

Alexendoo replied:
As would a base case without a binary expression

```rust
if if true { true } else { false } {}
```

If other lints get triggered by this add them to an `#![allow]`

Alexendoo commented on clippy_lints/src/stacked_ifs.rs:
```diff
@@ -0,0 +1,78 @@
+
+impl LateLintPass<'_> for StackedIfs {
+    fn check_expr(&mut self, cx: &LateContext<'_>, expr: &Expr<'_>) {
+        // Ensure that the expression isn't a part of a constant or macro expansion.
+        if expr.span.from_expansion() || in_constant(cx, expr.hir_id) {
```
The `in_constant` check can be removed as we do want to lint this inside `const` contexts

Alexendoo commented on clippy_lints/src/stacked_ifs.rs:
```diff
@@ -0,0 +1,78 @@
+        return;
+    };
+
+    // Remove any DropTemps wrapping the `if` statement.
+    let condition = match condition.kind {
+        ExprKind::DropTemps(expr) => expr,
+        _ => condition,
+    };
+
+    let condition = match condition.kind {
```
```suggestion
    let condition = match condition.peel_drop_temps().kind {
```

Alexendoo commented on clippy_lints/src/stacked_ifs.rs:
```diff
@@ -0,0 +1,78 @@
+        _ => condition,
+    };
+
+    if let ExprKind::If(_, _, _) = condition.kind {
```
Another macro check would be good here - `&& !condition.span.from_expansion()`

Macros that expand to `if` conditions aren't syntactically confusing so we don't need to lint them, a corresponding test would be

```rust
macro_rules! m {
    () => { if true { true } else { false } }
}

if m!() {}
```

ARandomDev99 commented:
I have also modified the code for the Binary expression part to now check if statements in both the left-hand side and right-hand side and added an additional test accordingly.

Alexendoo commented on clippy_lints/src/stacked_ifs.rs:
```diff
@@ -0,0 +1,94 @@
+        expr.span,
+        "stacked `if` found",
+        None,
+        "avoid using an `if` statement as a condition for another `if` statement",
```
`if` is an expression rather than a statement, a few other mentions of statement elsewhere too

Alexendoo commented on clippy_lints/src/stacked_ifs.rs:
```diff
@@ -0,0 +1,94 @@
+
+impl LateLintPass<'_> for StackedIfs {
+    fn check_expr(&mut self, cx: &LateContext<'_>, expr: &Expr<'_>) {
+        // Ensure that the expression isn't a part of a constant or macro expansion.
```
Comment is outdated

Alexendoo commented on tests/ui/stacked_ifs.rs:
```diff
@@ -0,0 +1,92 @@
+#![warn(clippy::stacked_ifs)]
+#![allow(clippy::manual_unwrap_or_default)]
+
+macro_rules! m {
+    () => {
+        let a = true;
+        let b = false;
+        if if a { b } else { true } {
+            println!("True");
+        }
+    };
+}
```
This is good to test the first `from_expansion` check, it would be good to have another that tests the `from_expansion` check of the condition like https://github.com/rust-lang/rust-clippy/pull/12548#discussion_r1536845445

Alexendoo commented:
Pending discussion here: https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy/topic/New.20lint.20process

ARandomDev99 commented:
I'll open a new PR incorporating the changes from https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy/topic/New.20lint.20process
